### PR TITLE
Issue #3370849 by nkoporec: Can't add an image to the existing album on stream page

### PR DESCRIPTION
**Pull request by nkoporec**
## Problem
When using social_album and I want to add an image to stream using the default /stream page, I see that a field widget is displayed to also add this image to an album, but the options there are always "None" or "Create a new album", even if I already have multiple albums created. So it should be possible to select an existing album.

I have multiple public albums created:

I don't see them in the field widget:

## Solution
We need to update the custom field widget to allow selecting an existing albums.

## Issue tracker
https://www.drupal.org/project/social/issues/3370849

## Theme issue tracker
*[Required if applicable] Paste a link to the drupal.org theme issue queue item, either from [socialbase](https://www.drupal.org/project/socialbase) or [socialblue](https://www.drupal.org/project/socialblue). If any other issue trackers were used, include links to those too.*

## How to test
1. Install social_album
2. Create a new public album
3. Go to /stream page and click on 'Add images'
4. Add an image, you should see the `Add image(s) to an album` field widget
5. Try to select an album that you created in step 2.

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] New features or changes to existing features are covered by tests, either unit (preferably) or behat
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Screenshots

Before:

After:

image

## Release notes
Added the ability to add images directly to existing albums via the stream page.

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  /**
+   * Fill out the album creation form and submit.
+   *
+   * Example: When I create a album using its creation page:
+   *              | Title       | My Album  |
+   *              | Description | It's the best |
+   *
+   * @When /^(?:|I )create a album using its creation page:$/
+   */
+  public function whenICreateABookUsingTheForm(TableNode $fields): void {
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:78]: Change whenICreateA**Book**UsingTheForm to whenICreate**AnAlbum**UsingTheForm

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  public function thenIShouldSeeTheAlbumIJustCreated() : void {
+    $fields = end($this->created);
+
+    $this->minkContext->assertPageContainsText("Album {$fields['title']} is successfully created. Now you can add images to this album.");
+  }
+
+  /**
+   * Fill out the album creation form and submit.
+   *
+   * Example: When I create a album using its creation page:
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:72]: Typo `a album` => `an album`

File: tests/behat/features/capabilities/post/post-album-existing.feature [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+  Scenario: Successfully add an image to the album via the stream post form
+
+    Given I enable the module "social_album"
+    Given users:
+      | name     | status | pass     | roles    |
+      | PostUser |      1 | PostUser | verified |
+    And I am logged in as "PostUser"
+
+    When I create a album using its creation page:
```
Comment (zanvidmar) [inline tests/behat/features/capabilities/post/post-album-existing.feature:15]: Typo `a album` => `an album`

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Drupal\social\Behat;
+
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Behat\Gherkin\Node\TableNode;
+use Behat\Mink\Exception\ElementNotFoundException;
+use Behat\MinkExtension\Context\RawMinkContext;
+use Drupal\Component\Plugin\Exception\PluginNotFoundException;
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:9]: PluginNotFoundException is not used. Please remove it.

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+use Drupal\DrupalExtension\Context\DrupalContext;
+use Drupal\DrupalExtension\Context\MinkContext;
+use Drupal\group\Entity\Group;
+use Drupal\node\Entity\Node;
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:13]: Drupal\group\Entity\Group and Drupal\node\Entity\Node are not used. Please remove them.

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        if ($fieldset === NULL) {
+          throw new ElementNotFoundException($this->getSession()->getDriver(), "fieldset", "named", "Basic Information");
+        }
+
+        $fieldset->fillField($field, $value);
+      }
+      // For the description we're using CKEditor so we must fill in the editor
+      // rather than the hidden form field.
+      // @todo Not being able to click the label shows an a11y issue.
+      elseif ($key === "description") {
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:103]: There is no "description" in Album.

**Review by zanvidmar — CHANGES_REQUESTED** [review]
Please check my comments in code. Thank you 🙂

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * Keep track of what we created, so we can use them for validating.
+   */
+  private array $created = [];
+
+  /**
+   * Provide help filling in the WYSIWYG editor.
+   */
+  private CKEditorContext $cKEditorContext;
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:30]: Do you still want to keep CKEditorContext, while this is not used at the moment?

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * Make some contexts available here so we can delegate steps.
+   *
+   * @BeforeScenario
+   */
+  public function gatherContexts(BeforeScenarioScope $scope) {
+    $environment = $scope->getEnvironment();
+
+    $this->cKEditorContext = $environment->getContext(CKEditorContext::class);
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:50]: Do you still want to keep CKEditorContext, while this is not used at the moment?

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+        $fieldset->fillField($field, $value);
+      }
+      elseif ($key === "group") {
+        $group_id = $this->getNewestGroupIdFromTitle($value);
+        if ($group_id === NULL) {
+          throw new \Exception("Group '${$value}' does not exist.");
+        }
+        $page->selectFieldOption($field, $group_id);
+        // Changing the group of a album updates the visibility settings so we
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:104]: Typo: a album => an album

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        $this->minkContext->iWaitForAjaxToFinish();
+      }
+      elseif ($key === "visibility") {
+        $page->selectFieldOption("field_content_visibility", $value);
+      }
+      else {
+        $page->fillField($field, $value);
+      }
+
+      // Changing the group of a album updates the visibility settings so we must
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:115]: Typo: a album => an album

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 11]
```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Drupal\social\Behat;
+
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Behat\Gherkin\Node\TableNode;
+use Behat\Mink\Exception\ElementNotFoundException;
+use Behat\MinkExtension\Context\RawMinkContext;
+use Drupal\DrupalExtension\Context\DrupalContext;
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:9]: Do you still want to keep DrupalContext, while this is not used at the moment?

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+   * Make some contexts available here so we can delegate steps.
+   *
+   * @BeforeScenario
+   */
+  public function gatherContexts(BeforeScenarioScope $scope) {
+    $environment = $scope->getEnvironment();
+
+    $this->cKEditorContext = $environment->getContext(CKEditorContext::class);
+    $this->minkContext = $environment->getContext(SocialMinkContext::class);
+    $this->drupalContext = $environment->getContext(SocialDrupalContext::class);
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:52]: Do you still want to keep DrupalContext, while this is not used at the moment?

File: tests/behat/features/bootstrap/AlbumContext.php [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * The Drupal mink context is useful for validation of content.
+   */
+  private MinkContext $minkContext;
+
+  /**
+   * The Drupal context which gives us access to user management.
+   */
+  private DrupalContext $drupalContext;
```
Comment (zanvidmar) [inline tests/behat/features/bootstrap/AlbumContext.php:40]: Do you still want to keep DrupalContext, while this is not used at the moment?

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by BiaInacio** [comment]
Change is working as expected!
image

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Review by zanvidmar — APPROVED** [review]

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by Kingdutch** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3442-n1o2bgtglsebhagxwlbk8ympvrmkfnkc.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/649bd517757563ca371eccda

**Comment by zanvidmar** [comment]
🍒 cherry picked

Cherry-picked to 11.8.x here: [0c2e368fc](https://github.com/goalgorilla/open_social/commit/0c2e368fc4ed033a53695c774c236164e3158886)
Cherry-picked to 11.9.x here: [655c43dca](https://github.com/goalgorilla/open_social/commit/655c43dcab7674c8fa874bef8412273e8f769c40)

**Pull request merged**